packages/sdk/src/oracle/services/statemachines/statemachine.ts

```diff
@@ -0,0 +1,106 @@
+import { Context, Memory, ContextType } from "../../types/statemachine";
+import type Store from "../../store";
+import { ContextManager } from "./utils";
+
+// context types, import new ones here
+import * as setUser from "./setUser";
+import * as setActiveRequest from "./setActiveRequest";
+
+/**
+ * StateMachine. This class will be used to handle all change requests by the user, including setting state which
+ * may require triggering fetching data from chain, or initiating transactions that require tracking.
+ *
+ * This class is meant to step through states of a Context object. This object can have variable
+ * parameters and memory. There are several concepts to understand before using this class:
+ *
+ * 1. Handlers - This is the state machine state handlers, it is an objected keyed by each state, with a function handler.
+ * 2. Params - These are common parameters set by the caller, passed into each handler function.
+ * 3. Memory - This is a writing space within the state machine handlers that accumulates through states.
+ * 4. Context - This is the sum of all data needed to transition through the state machine states. It includes params, memory and metadata about the runtime.
+ * 5. Context.done - This is a special property on context.done which represents that the state machine is done transitioning this context.
+ * 6. Context.state = "done" - This is a reserved state on context.state, if set to "done" its the same thing as done = true.
+ * 7. Context.state = "error" - This is a reserved state on context.state, if set to "error" it means the context is done, but also there is an context.error object.
+ * 8. Interval - This is a property on the context which specifies the maximum rate in MS this context will transition, based on the current timestamp.
+ */
+export class StateMachine {
+  private pending: Context<unknown, unknown & Memory>[] = [];
+  // this needs to be updated when adding new context type
+  public types: {
+    [ContextType.setUser]: ContextManager<setUser.Params, setUser.Memory>;
+    [ContextType.setActiveRequest]: ContextManager<setActiveRequest.Params, setActiveRequest.Memory>;
```

Comment:
adding new types is a really tedious process, theres no way to build up this object at run time because you dont have access to typescript types at that time. otherwise we could just add new context types to a list and dyanmically build the structure ( like you would in js). Now we have to manually add it here and a handful of other places in what i consider infrastructure.

Reply:
hmm yea i am open to any ideas here. with the current design theres a single queue of generic "contexts".  In order to create contexts and push onto queue we need to know the type to properly check parameters, and when popping off the queue and stepping through them we need to know the type.